src/sql/src/plan/query.rs

```diff
@@ -2156,14 +2188,87 @@ fn plan_rows_from(
         return plan_solitary_table_function(qcx, function, alias, with_ordinality);
     }
 
+    // Per PostgreSQL, all scope items take the name of the first function
+    // (unless aliased).
+    // See: https://github.com/postgres/postgres/blob/639a86e36/src/backend/parser/parse_relation.c#L1701-L1705
+    let (expr, mut scope, num_cols) =
+        plan_rows_from_internal(qcx, functions, Some(&functions[0].name))?;
+
+    // Columns tracks the set of columns we will keep in the projection.
+    let mut columns = Vec::new();
+    let mut offset = 0;
+    // Retain table function's non-ordinality columns.
+    for (idx, cols) in num_cols.into_iter().enumerate() {
+        for i in 0..cols {
+            columns.push(offset + i);
+        }
+        offset += cols + 1;
+
+        // Remove the ordinality column from the scope, accounting for previous scope
+        // changes from this loop.
+        scope.items.remove(offset - idx - 1);
+    }
+
+    // If `WITH ORDINALITY` was specified, include the coalesced ordinality
+    // column. Otherwise remove it from the scope.
+    if with_ordinality {
+        columns.push(scope.items.len());
+    } else {
+        scope.items.pop();
+    }
+
+    let expr = expr.project(columns);
+
+    let scope = plan_table_alias(scope, alias)?;
+    Ok((expr, scope))
+}
+
+/// Plans an expression coalescing multiple table functions. Each table
+/// function is followed by its row ordinality. The entire expression is
+/// followed by the coalesced row ordinality.
+///
+/// The returned Scope will set all item's table_name's to the `table_name`
+/// parameter if it is `Some`. If `None`, they will be the name of each table
+/// function.
+///
+/// The returned `Vec<usize>` is the number of (non-ordinality) columns from
+/// each table function.
+///
+/// For example, with table functions tf1 returning 1 column (a) and tf2
+/// returning 2 columns (b, c), this function will return an expr 6 columns:
+///
+/// - tf1.a
+/// - tf1.ordinality
+/// - tf2.b
+/// - tf2.c
+/// - tf2.ordinality
+/// - coalesced_ordinality
```

Comment:
including what the `Vec<usize>` would be would also help folks cement their understanding.